modules/core/arrow-annotations-processor/src/main/java/arrow/generic/ProductProcessor.kt

```diff
@@ -0,0 +1,84 @@
+package arrow.generic
+
+import com.google.auto.service.AutoService
+import arrow.common.utils.AbstractProcessor
+import arrow.common.utils.asClassOrPackageDataWrapper
+import arrow.common.utils.knownError
+import me.eugeniomarletti.kotlin.metadata.KotlinClassMetadata
+import me.eugeniomarletti.kotlin.metadata.extractFullName
+import me.eugeniomarletti.kotlin.metadata.isDataClass
+import me.eugeniomarletti.kotlin.metadata.kotlinMetadata
+import me.eugeniomarletti.kotlin.metadata.proto
+import org.jetbrains.kotlin.serialization.ProtoBuf
+import java.io.File
+import javax.annotation.processing.Processor
+import javax.annotation.processing.RoundEnvironment
+import javax.lang.model.SourceVersion
+import javax.lang.model.element.TypeElement
+
+import javax.lang.model.element.Element
+
+@AutoService(Processor::class)
+class ProductProcessor : AbstractProcessor() {
+
+    private val annotatedProduct = mutableListOf<AnnotatedGeneric>()
+
+    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()
+
+    override fun getSupportedAnnotationTypes() = setOf(productAnnotationClass.canonicalName)
+
+    override fun onProcess(annotations: Set<TypeElement>, roundEnv: RoundEnvironment) {
+        annotatedProduct += roundEnv
+                .getElementsAnnotatedWith(productAnnotationClass)
+                .map(this::evalAnnotatedProductElement)
+
+        if (roundEnv.processingOver()) {
+            val generatedDir = File(this.generatedDir!!, "").also { it.mkdirs() }
+            ProductFileGenerator(annotatedProduct, generatedDir).generate()
+        }
+    }
+
+    private fun productAnnotationError(element: Element, annotationName: String, targetName: String): String = """
+            |Cannot use $annotationName on ${element.enclosingElement}.${element.simpleName}.
+            |It can only be used on $targetName.""".trimMargin()
+
+    private fun evalAnnotatedProductElement(element: Element): AnnotatedGeneric = when {
+        (element.kotlinMetadata as? KotlinClassMetadata)?.data?.classProto?.isDataClass == true -> {
+            val elementClassData = getClassData(element)
+            val paramNames = getConstructorParamNames(element)
+            val typeNames = getConstructorTypesNames(element)
+            val properties = paramNames.zip(typeNames).map { Target(it.second,it.first) }
+            if (properties.size > 22)
+                knownError("${element.enclosingElement}.${element.simpleName} up to 22 constructor parameters is supported")
```

Comment:
Reading the `knownError()` implementation seems like you're throwing exceptions when something goes wrong. Why aren't you using the procesing environment `messager` for these error logs?

Messager plays well with compile time testing frameworks like `compile-testing` and `Truth` which allow you to assert over generated classes, errors, etc. [Some test samples here](https://github.com/JorgeCastilloPrz/corleone/blob/master/corleone/src/test/java/com/github/jorgecastilloprz/corleone/AnnotationValidationTest.java).

I'm not sure about how problematic can become to just fail by throwing.

Reply:
Sounds good, but that was already there and other processors may depend on it so it should be tackle in a different issue. Hopefully `kotlin-meta` will make meta programming easier once we have an AST we control ourselves, the current state of meta-programming facilities in these processor is far from ideal and the proto APIs a royal pain in the ass to deal with.

Reply:
OK #766

Reply:
Throwing works fine with tests:
https://github.com/arrow-kt/arrow/blob/master/modules/core/arrow-annotations-processor-test/src/test/java/arrow/ap/tests/OptionalTest.kt#L20
https://github.com/arrow-kt/arrow/blob/master/modules/core/arrow-annotations-processor-test/src/test/java/arrow/ap/tests/APTest.kt#L48

But printing error with messager automatically throws exception (https://docs.oracle.com/javase/7/docs/api/javax/annotation/processing/Messager.html), so I guess that's the correct way of reporting errors.